### ISSUE TITLE
Having non-nullary identifiers in a pattern match succeeds

### ISSUE THREAD

**Issue by LiamGoodacre** [idris-lang/Idris-dev#2351]
Pattern matching using identifiers that refer to non-nullary constructions doesn't fail.

The following introduces `Just`, `Maybe`, and `example` as pattern bindings.

``` idris
example : Unit -> Unit -> Unit -> (Unit, Unit, Unit)
example Just Maybe example = (Just, Maybe, example)
```

If you use a nullary construction (`Nothing`):

``` idris
example' : Unit -> Unit
example' Nothing = ()
```

It does error:

```
When elaborating left hand side of example':
When elaborating an application of Main.example':
        Can't unify
                Maybe a (Type of Nothing)
        with
                () (Expected type)
```

**Comment by edwinb** [comment]
Yes, it's treating it as a pattern variable because it doesn't work as a
constructor. I'd expect this behaviour, but that doesn't mean this
behaviour is the right one.

I have thought for a while that this is strange. At first, I was trying
to avoid any necessary distinction between things beginning with lower
case letters, and things beginning with upper case, but this leads to
various weird rules and problems like this.

Perhaps we should apply the same rule we do in types, specifically: only
names which begin with a lower case letter will be bound in pattern
clauses, and anything else must already be bound globally.

**Comment by david-christiansen** [comment]
Scala has this rule for patterns, and it seems to work fine there.

**Comment by ahmadsalim** [comment]
I think we should explicit have the rule that pattern variables use lowercase letters, to avoid confusion.

**Comment by supergrade** [comment]
Lowercase restriction, as you suggest, is probably the more consistent solution. But in another sphere I have 'unused variable' warnings on pattern variables also tend to catch these.

**Comment by ahmadsalim** [comment]
Fixed by db690778858de4b756f3b64a11763aa825574ea0
